src/main/java/io/confluent/copycat/hdfs/HdfsWriter.java

```diff
@@ -1,167 +1,292 @@
   private Map<TopicPartition, Long> offsets = null;
   private Map<TopicPartition, Integer> recordCounters = null;
   private Configuration conf;
-  private FileSystem fs;
-  private Class<? extends RecordWriterProvider> writerProviderClass;
+  private RecordWriterProvider writerProvider;
   private String url;
   private int flushSize;
   private String topicsDir;
+  private Set<TopicPartition> assignment;
+  private AvroData avroData;
+  private SinkTaskContext context;
+  private Map<TopicPartition, State> states;
+  private Set<TopicPartition> recovered;
+  private Map<TopicPartition, Queue<SinkRecord>> buffer;
+  private Storage storage;
+  private long backOffMs;
+
+  private enum State {
+    RECOVERY_STARTED,
+    RECOVERY_PARTITION_PAUSED,
+    WAL_CREATED,
+    WAL_APPLIED,
+    WAL_TRUNCATED,
+    OFFSET_RESET,
+    WRITE_STARTED,
+    WRITE_PARTITION_PAUSED,
+    SHOULD_ROTATE,
+    TEMP_FILE_CLOSED,
+    WAL_APPENDED,
+    FILE_COMMITTED;
+
+    private static State[] vals = values();
+    public State next() {
+      return vals[(this.ordinal() + 1) % vals.length];
+    }
+  }
+
 
   @SuppressWarnings("unchecked")
-  public HdfsWriter(HdfsSinkConnectorConfig connectorConfig) {
+  public HdfsWriter(HdfsSinkConnectorConfig connectorConfig, SinkTaskContext context, AvroData avroData) {
     try {
-      writers = new HashMap<TopicPartition, RecordWriter<Long, SinkRecord>>();
-      recordCounters = new HashMap<TopicPartition, Integer>();
-      flushSize = connectorConfig.getInt(HdfsSinkConnectorConfig.FLUSH_SIZE_CONFIG);
+      writers = new HashMap<>();
+      wals = new HashMap<>();
+      offsets = new HashMap<>();
+      recordCounters = new HashMap<>();
+      tempFileNames = new HashMap<>();
 
+      flushSize = connectorConfig.getInt(HdfsSinkConnectorConfig.FLUSH_SIZE_CONFIG);
+      backOffMs = connectorConfig.getLong(HdfsSinkConnectorConfig.RETRY_BACKOFF_CONFIG);
       url = connectorConfig.getString(HdfsSinkConnectorConfig.HDFS_URL_CONFIG);
       topicsDir = connectorConfig.getString(HdfsSinkConnectorConfig.TOPIC_DIR_CONFIG);
       conf = new Configuration();
-      fs = FileSystem.newInstance(URI.create(url), conf);
-      Path path = new Path(url + "/" + topicsDir);
-      if (!fs.exists(path)) {
-        fs.mkdirs(path);
+
+      Class<? extends Storage> storageClass =
+          (Class<? extends Storage>) Class.forName(connectorConfig.getString(HdfsSinkConnectorConfig.STORAGE_CLASS_CONFIG));
+      this.storage = StorageFactory.createStorage(storageClass, conf, url);
+      createTopicsDir();
+      writerProvider = ((Class<RecordWriterProvider>) Class.forName(connectorConfig.getString(
+              HdfsSinkConnectorConfig.RECORD_WRITER_PROVIDER_CLASS_CONFIG))).newInstance();
+
+      this.context = context;
+      this.assignment = context.assignment();
+      this.avroData = avroData;
+
+      states = new HashMap<>();
+      for (TopicPartition tp: assignment) {
+        states.put(tp, State.RECOVERY_STARTED);
```

Comment:
Can (or should) we pause the partitions here? Maybe this ends up not mattering, but it seems weird to potentially get some records on the first `put()` which you add to the buffer and then ignore when you didn't even really want them yet. Then again, we have to handle those uncommitted messages in the buffer, so maybe it makes no difference.